mod_form.php

```diff
@@ -55,33 +55,80 @@ public function definition() {
 
         $this->standard_intro_elements(get_string('description', 'customcert'));
 
-        $mform->addElement('header', 'options', get_string('options', 'customcert'));
+        $optionsheader = $mform->createElement('header', 'options', get_string('options', 'customcert'));
 
-        $mform->addElement('selectyesno', 'emailstudents', get_string('emailstudents', 'customcert'));
-        $mform->setType('emailstudents', 0);
-        $mform->addHelpButton('emailstudents', 'emailstudents', 'customcert');
+        if (has_capability('mod/customcert:manageemailstudents', $this->get_context())) {
+            $mform->addElement('selectyesno', 'emailstudents', get_string('emailstudents', 'customcert'));
+            $mform->setDefault('emailstudents', get_config('customcert', 'emailstudents'));
+            $mform->addHelpButton('emailstudents', 'emailstudents', 'customcert');
+            $firstoption = 'emailstudents';
+        } else {
+            $mform->addElement('hidden', 'emailstudents', get_config('customcert', 'emailstudents'));
```

Comment:
I am wondering if adding the setting as hidden element with the default value is correct in all cases.

Imagine this chain of actions:
* User A has mod/customcert:manageemailstudents, user B has not
* User A adds a new instance of mod_customcert to the course and sets this setting to a value which differs from the default.
* User B edits the existing instance and saves the form.
* Then, if I am not wrong, User B will have changed the setting back to the default although he does not have the capability to change the setting.

Do we need to differentiate if the instance will be created or has already been created before and, in the second case, set the hidden element to the existing instance setting instead of to the default?

Reply:
Very good point! Will fix it.

Reply:
Thanks for confirming my thoughts :)
To make it really bullet-proof, you could also check the capabilities again within data_postprocessing()

Reply:
Actually I have just tested it and looks like all works good. Default values get saved just for User B when he created a new activity.